### PR TITLE
Support behavioral changes to instance, instance template's max_run_duration feature on Beta

### PR DESCRIPTION
**Pull request by arungantasala**
Max run duration(MRD) feature support in terraform was added in https://github.com/GoogleCloudPlatform/magic-modules/pull/6812. There are behavioral changes in the feature which needs further changes.

1) max_run_duration is not a one time intent
2) Add STOP instance termination action with MRD
3) Support MRD on instances with attached local SSDs.

See SchedulingOnInstanceStopAction section in https://pkg.go.dev/google.golang.org/api/compute/v0.beta.

Fix for https://github.com/hashicorp/terraform-provider-google/issues/13005

Closes https://github.com/hashicorp/terraform-provider-google/issues/13671

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added `on_instance_stop_action` field to `google_compute_instance`, `google_compute_instance_template`, and `google_compute_instance_from_machine_image ` resources (beta)
```

### CODE REVIEW

**Review by SarahFrench — CHANGES_REQUESTED** [review]
Thanks for your PR - the new tests look good and are passing but I want to make sure [this message](https://github.com/GoogleCloudPlatform/magic-modules/pull/10683#issuecomment-2108474157) about missing test coverage is acknowledged.

That message shows that the new `on_instance_stop_action` field is also being added to the `google_compute_instance_from_machine_image` resource. This is because schemas are re-used in the provider ([see here](https://github.com/hashicorp/terraform-provider-google-beta/blob/3d8fce747cf855191df3e92cc224121d90a4abcb/google-beta/services/compute/resource_compute_instance_from_machine_image.go#L37) for google_compute_instance_from_machine_image).

Firstly, could you please update the documentation for google_compute_instance_from_machine_image to include the new field?

And then, could you please add a basic acceptance test  that tests the new field/behaviour you've added to `google_compute_instance_from_machine_image`?

I'm happy for us to ignore [the message](https://github.com/GoogleCloudPlatform/magic-modules/pull/10683#issuecomment-2108474157)'s warnings about lack of test coverage on fields that your PR isn't adding.

**Comment by arungantasala** [comment]
1) `google_compute_instance_from_machine_image` [documentation](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance_from_machine_image) mentions "In addition to these, most* arguments from `google_compute_instance` are supported as a way to override the properties in the machine image."  This includes the new argument `on_instance_stop_action`
2) Added acceptance tests for the new field at all the relavant places

**Review by SarahFrench — COMMENTED** [review]
The [problem reported here](https://github.com/GoogleCloudPlatform/magic-modules/pull/10683#issuecomment-2113963337) is a panic that occurs when running the `TestAccComputeInstanceNetworkIntefaceWithSecurityPolicy/access_config_update_access_config` acceptance test:

```
panic: interface conversion: interface {} is []interface {}, not map[string]interface {}

goroutine 68179 [running]:
github.com/hashicorp/terraform-provider-google-beta/google-beta/services/compute.hasMaxRunDurationChanged(...)
	/go/src/github.com/modular-magician/terraform-provider-google-beta/google-beta/services/compute/compute_instance_helpers.go:705
github.com/hashicorp/terraform-provider-google-beta/google-beta/services/compute.schedulingHasChangeRequiringReboot(0xc001302ca8?)
	/go/src/github.com/modular-magician/terraform-provider-google-beta/google-beta/services/compute/compute_instance_helpers.go:658 +0x3ce
github.com/hashicorp/terraform-provider-google-beta/google-beta/services/compute.resourceComputeInstanceUpdate(0xc001f3f680, {0x4613dc0?, 0xc000161000?})
	/go/src/github.com/modular-magician/terraform-provider-google-beta/google-beta/services/compute/resource_compute_instance.go:1890 +0x1437
```

The [line compute_instance_helpers.go:705 in the generated provider](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10683-old..auto-pr-10683#diff-67d4ccf0a83752cab376f26e348528f8712234a6ba2e4b7805b96f459283b665R705) is the beginning of hasMaxRunDurationChanged function where empty interfaces are converted to maps

**Review by SarahFrench — CHANGES_REQUESTED** [review]
Please address the panic described in https://github.com/GoogleCloudPlatform/magic-modules/pull/10683#pullrequestreview-2060426595

**Comment by arungantasala** [comment]
Thanks Sarah for pointing this out. I was not able to reproduce the issue locally by running the test you pointed out. It is somehow failing on its own due to an unrelated issue. But, i have pushed a change to fix this. I will take another look if the panic shows up again.

**Comment by arungantasala** [comment]
Hi Sarah,
The panic error is now fixed. Could you please trigger the checks again and review this PR? We are planning to make this available to the customers in the next release.
Thanks.

**Review by SarahFrench — DISMISSED** [review]
Looks good - the panic is addressed and there is test coverage for all the new fields added by this PR.

File: mmv1/third_party/terraform/services/compute/compute_instance_helpers.go.erb [hunk 1]
```diff
@@ -656,7 +692,7 @@ func schedulingHasChangeRequiringReboot(d *schema.ResourceData) bool {
 	oScheduling := o.([]interface{})[0].(map[string]interface{})
 	newScheduling := n.([]interface{})[0].(map[string]interface{})
 
-	return hasNodeAffinitiesChanged(oScheduling, newScheduling)
+	return hasNodeAffinitiesChanged(oScheduling, newScheduling) || hasMaxRunDurationChanged(oScheduling, newScheduling)
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/services/compute/compute_instance_helpers.go.erb:695]: This suggested change should mean that the changes only appear in the Beta provider

```suggestion
<% unless version == 'ga' -%>
	return hasNodeAffinitiesChanged(oScheduling, newScheduling) || hasMaxRunDurationChanged(oScheduling, newScheduling)
<% else -%>
	return hasNodeAffinitiesChanged(oScheduling, newScheduling)
<% end -%>
```

File: mmv1/third_party/terraform/services/compute/resource_compute_instance.go.erb [hunk 2]
```diff
@@ -2339,7 +2348,7 @@ func resourceComputeInstanceUpdate(d *schema.ResourceData, meta interface{}) err
 		desiredStatus := d.Get("desired_status").(string)
 
 		if statusBeforeUpdate == "RUNNING" && desiredStatus != "TERMINATED" && !d.Get("allow_stopping_for_update").(bool) {
-			return fmt.Errorf("Changing the machine_type, min_cpu_platform, service_account, enable_display, shielded_instance_config, scheduling.node_affinities " +
+			return fmt.Errorf("Changing the machine_type, min_cpu_platform, service_account, enable_display, shielded_instance_config, scheduling.node_affinities, scheduling.max_run_duration " +
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/services/compute/resource_compute_instance.go.erb:2351]:
```suggestion
<% unless version == 'ga' -%>
			return fmt.Errorf("Changing the machine_type, min_cpu_platform, service_account, enable_display, shielded_instance_config, scheduling.node_affinities, scheduling.max_run_duration " +
<% else -%>
			return fmt.Errorf("Changing the machine_type, min_cpu_platform, service_account, enable_display, shielded_instance_config, scheduling.node_affinities " +
<% end -%>
```

**Comment by arungantasala** [comment]
Thanks Sarah for catching this. I have made changes as you suggested.

**Review by SarahFrench — APPROVED** [review]
Thanks for addressing the GA/Beta issue! I can see that the diffs for the GA provider still show changes in functions like `testAccComputeInstanceTemplate_spot_maxRunDuration`, however those functions are unused in the GA code base so _the code change isn't user-facing_. Because Go tolerates unused functions and those functions were already present in the GA code before this PR, I'm happy to merge this PR

**Pull request merged**